src/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs

```diff
@@ -0,0 +1,631 @@
+        }
+
+        [Fact]
+        public void CopyToAsync_MockContentReturnsNull_ThrowsInvalidOperationException()
+        {
+            // return 'null' when CopyToAsync() is called.
+            var content = new MockContent(MockOptions.ReturnNullInCopyToAsync); 
+            var m = new MemoryStream();
+            
+            // The HttpContent derived class (MockContent in our case) must return a Task object when WriteToAsync()
+            // is called. If not, HttpContent will throw.
+            Assert.Throws<InvalidOperationException>(() => { Task t = content.CopyToAsync(m); });
+        }
+
+        [Fact]
+        public async Task CopyToAsync_BufferContentFirst_UseBufferedStreamAsSource()
+        {
+            var data = new byte[10];
+            var content = new MockContent(data);
+            content.LoadIntoBufferAsync().Wait();
+
+            Assert.Equal(1, content.SerializeToStreamAsyncCount);
+            var destination = new MemoryStream();
+            await content.CopyToAsync(destination);
+
+            // Our MockContent should not be called for the CopyTo() operation since the buffered stream should be 
+            // used.
+            Assert.Equal(1, content.SerializeToStreamAsyncCount);
+            Assert.Equal(data.Length, destination.Length);
+        }
+
+        [Fact]
+        public void TryComputeLength_RetrieveContentLength_ComputeLengthShouldBeCalled()
+        {
+            var content = new MockContent(MockOptions.CanCalculateLength);
+
+            Assert.Equal(content.GetMockData().Length, content.Headers.ContentLength);
+            Assert.Equal(1, content.TryComputeLengthCount);
+        }
+
+        [Fact]
+        public async Task TryComputeLength_RetrieveContentLengthFromBufferedContent_ComputeLengthIsNotCalled()
+        {
+            var content = new MockContent();
+            await content.LoadIntoBufferAsync();
+
+            Assert.Equal(content.GetMockData().Length, content.Headers.ContentLength);
+            
+            // Called once to determine the size of the buffer.
+            Assert.Equal(1, content.TryComputeLengthCount); 
+        }
+
+        [Fact]
+        public void TryComputeLength_ThrowCustomExceptionInOverriddenMethod_ExceptionBubblesUpToCaller()
+        {
+            var content = new MockContent(MockOptions.ThrowInTryComputeLength); 
+
+            var m = new MemoryStream();
+            Assert.Throws<MockException>(() => content.Headers.ContentLength);
+        }
+
+        [Fact]
+        public async Task ReadAsStreamAsync_GetFromUnbufferedContent_CreateContentReadStreamCalledOnce()
+        {
+            var content = new MockContent(MockOptions.CanCalculateLength);
+
+            // Call multiple times: CreateContentReadStreamAsync() should be called only once.
+            Stream stream = await content.ReadAsStreamAsync();
+            stream = await content.ReadAsStreamAsync();
+            stream = await content.ReadAsStreamAsync();
```

Comment:
Do you want to Assert.Same that the streams you get back are the same instance each time? (assuming they're meant to be)